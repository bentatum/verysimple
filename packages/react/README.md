# @verysimple/react

This is the `@verysimple/react` package