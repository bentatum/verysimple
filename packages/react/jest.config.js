/** @type {import('jest').Config} */
module.exports = {
  testEnvironment: "jsdom",
  setupFilesAfterEnv: ["@testing-library/jest-dom/extend-expect"],
  collectCoverage: true,
  coverageThreshold: {
    global: {
      statements: 75,
      branches: 72,
      functions: 75,
      lines: 75,
    },
  },
  coverageReporters: ["text-summary", "json-summary"],
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!**/*.d.ts"],
  coverageDirectory: "coverage",
  moduleNameMapper: {
    "@/test/(.*)": "<rootDir>/test/$1",
    "@/(.*)$": "<rootDir>/src/$1",
  },
  modulePathIgnorePatterns: ["<rootDir>/dist/"],
  prettierPath: null,
};
