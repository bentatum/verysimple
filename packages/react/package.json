{
  "name": "@verysimple/react",
  "version": "8.1.1",
  "description": "A very simple component library",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "author": "Ben Tatum <mail@benjtatum.com>",
  "license": "MIT",
  "repository": "https://github.com/bentatum/verysimple",
  "scripts": {
    "test": "jest",
    "prebuild": "rm -rf dist",
    "build": "tsc --project tsconfig.build.json && tsc-alias"
  },
  "peerDependencies": {
    "react": ">=16",
    "react-dom": ">=16"
  },
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@testing-library/dom": "^8.20.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/inquirer": "latest",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.21",
    "@types/testing-library__jest-dom": "^5.14.5",
    "babel-jest": "^29.0.3",
    "babel-loader": "^8.2.5",
    "classnames": "^2.3.2",
    "framer-motion": "^10.15.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwindcss": "^3.1.8",
    "tsc-alias": "^1.8.2",
    "typescript": "^4.9.3"
  },
  "volta": {
    "node": "20.10.0",
    "yarn": "1.22.21"
  },
  "dependencies": {
    "@headlessui/react": "^2.1.9",
    "@heroicons/react": "^2.0.13",
    "clsx": "^2.1.0",
    "framer-motion": "^10.15.0"
  }
}
