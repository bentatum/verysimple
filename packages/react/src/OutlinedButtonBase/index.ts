export { default } from "./OutlinedButtonBase";
export * from "./OutlinedButtonBase";
