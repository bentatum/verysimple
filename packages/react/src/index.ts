export * from "./ButtonBase";
export * from "./Card";
export * from "./Dialog";
export * from "./FilledButtonBase";
export * from "./FilledIconButton";
export * from "./GradientBorder";
export * from "./helpers";
export * from "./IconButton";
export * from "./NormalIconButton";
export * from "./OutlinedButtonBase";
export * from "./OutlinedIconButton";
export * from "./Spinner";
export * from "./types";
