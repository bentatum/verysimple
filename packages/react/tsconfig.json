{
  "extends": "@verysimple/tsconfig/react-library.json",
  "compilerOptions": {
    "outDir": "./dist",
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["node_modules"]
}
