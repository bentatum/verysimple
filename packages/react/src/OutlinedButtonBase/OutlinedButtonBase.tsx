import classNames from "classnames";
import { FC, forwardRef } from "react";
import { ButtonBase, ButtonBaseProps } from "@/ButtonBase";
import { GradientBorder, GradientBorderProps } from "@/GradientBorder";
import type { ButtonColor } from "@/types";
import {
  buttonShadowClassNames,
  buttonTextClassNames,
  fieldSizeClassNames,
  fieldBorderRadiusClassNames,
} from "@/helpers";

export interface OutlinedButtonBaseProps extends ButtonBaseProps {
  color?: ButtonColor;
  GradientBorderProps?: GradientBorderProps;
  rounded?: boolean | "full";
}

export const OutlinedButtonBase: FC<OutlinedButtonBaseProps> = forwardRef<
  HTMLButtonElement,
  OutlinedButtonBaseProps
>(
  (
    {
      color = "neutral",
      className = "",
      disabled,
      size = "md",
      as = "button",
      rounded = "full",
      GradientBorderProps,
      ...props
    },
    ref
  ) => {
    const GradientBorderButton = forwardRef<any, GradientBorderProps>(
      (props, ref) => {
        return (
          <GradientBorder
            {...props}
            as={as}
            ref={ref}
            borderColor={disabled ? "var(--my-border-disabled)" : undefined}
            {...GradientBorderProps}
          />
        );
      }
    );

    const Component = color !== "primary" ? as : GradientBorderButton;
    const gradientBorderEligible =
      color !== "destructive" &&
      !disabled &&
      !className.match(/hover:text-|focus:text-/);

    return (
      <ButtonBase
        {...props}
        ref={ref}
        size={size}
        as={Component}
        disabled={disabled}
        className={classNames(
          // this keeps the size consistent with inputs
          'box-content',
          buttonShadowClassNames("outlined", className),
          buttonTextClassNames(color, "outlined", className),
          fieldBorderRadiusClassNames(rounded, className),
          fieldSizeClassNames(size),
          {
            border: color !== "primary",
            "my-border-primary": color === "neutral",
            "border-red-500 disabled:border-red-400": color === "destructive",
            "hover:my-bg-gradient-dark-to-r hover:border-none hover:p-[1px] hover:text-white":
              gradientBorderEligible,
            "focus:my-bg-gradient-dark-to-r focus:border-none focus:p-[1px] focus:text-white":
              gradientBorderEligible,
          },
          className
        )}
      />
    );
  }
);

OutlinedButtonBase.displayName = "OutlinedButtonBase";

export default OutlinedButtonBase;
