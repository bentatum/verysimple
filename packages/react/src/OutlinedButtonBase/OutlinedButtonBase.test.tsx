import { render, screen } from "@testing-library/react";
import { createRef } from "react";
import OutlinedButtonBase from "./OutlinedButtonBase";

describe("OutlinedButtonBase", () => {
  it("should have an outline", () => {
    const { container } = render(<OutlinedButtonBase>Test</OutlinedButtonBase>);
    expect(screen.getByRole("button")).toHaveClass("border");
    expect(container).toMatchSnapshot();
  });
  it("should have normal disabled styled when destructive", () => {
    render(
      <OutlinedButtonBase color="destructive" disabled>
        Test
      </OutlinedButtonBase>
    );
    expect(screen.getByRole("button")).toHaveClass(
      "border-red-500",
      "disabled:border-red-400"
    );
  });

  it("forwards the ref", () => {
    const ref = createRef<HTMLButtonElement>();
    render(<OutlinedButtonBase ref={ref}>test</OutlinedButtonBase>);
    expect(ref.current).toBeInTheDocument();
  });
  it("has a display name", () => {
    expect(OutlinedButtonBase.displayName).toBe("OutlinedButtonBase");
  });
});
