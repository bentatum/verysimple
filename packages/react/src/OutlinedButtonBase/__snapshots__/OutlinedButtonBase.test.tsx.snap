// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutlinedButtonBase should have an outline 1`] = `
<div>
  <button
    class="focus:outline-none cursor-pointer disabled:cursor-not-allowed inline-flex items-center justify-center rounded border shadow hover:shadow disabled:shadow-none transition-shadow ease-in duration-100 whitespace-nowrap font-medium text-base min-h-[44px] border box-border"
    tabindex="0"
  >
    Test
  </button>
</div>
`;
